metadata-ingestion/src/datahub/ingestion/source/delta_lake/source.py

```diff
@@ -286,45 +292,70 @@ def ingest_table(
-    ) -> Iterable[MetadataWorkUnit]:
+    @cached_property
+    def storage_options(self) -> Dict[str, str]:
+        if (
+            self.source_config.is_s3
+            and self.source_config.s3 is not None
+            and self.source_config.s3.aws_config is not None
+        ):
+            aws_config = self.source_config.s3.aws_config
+            creds = aws_config.get_credentials()
+            opts = {
+                "AWS_ACCESS_KEY_ID": creds.get("aws_access_key_id", ""),
+                "AWS_SECRET_ACCESS_KEY": creds.get("aws_secret_access_key", ""),
+                "AWS_SESSION_TOKEN": creds.get("aws_session_token", ""),
+                # Allow http connections, this is required for minio
+                "AWS_STORAGE_ALLOW_HTTP": "true",
+            }
+            if self.source_config.s3.aws_config.aws_region:
+                opts["AWS_REGION"] = aws_config.aws_region
+            if self.source_config.s3.aws_config.aws_endpoint_url:
+                opts["AWS_ENDPOINT_URL"] = aws_config.aws_endpoint_url
+            return opts
+        else:
+            return {}
+
+    def process_folder(self, path: str) -> Iterable[MetadataWorkUnit]:
         logger.debug(f"Processing folder: {path}")
-        delta_table = read_delta_table(path, self.source_config)
+        delta_table = read_delta_table(path, self.storage_options, self.source_config)
         if delta_table:
             logger.debug(f"Delta table found at: {path}")
             for wu in self.ingest_table(delta_table, path):
                 yield wu
         else:
-            for folder in get_folders(path):
-                yield from self.process_folder(path + "/" + folder, get_folders)
+            for folder in self.get_folders(path):
+                yield from self.process_folder(folder)
+
+    def get_folders(self, path: str):
+        if self.source_config.is_s3:
+            return self.s3_get_folders(path)
+        else:
+            return self.local_get_folders(path)
 
     def s3_get_folders(self, path: str) -> Iterable[str]:
-        if self.source_config.s3 is not None:
-            yield from list_folders_path(path, self.source_config.s3.aws_config)
+        parse_result = urlparse(path)
+        for page in self.s3_client.get_paginator("list_objects_v2").paginate(
+            Bucket=parse_result.netloc, Prefix=parse_result.path[1:], Delimiter="/"
+        ):
+            for o in page.get("CommonPrefixes", []):
+                yield f"{parse_result.scheme}://{parse_result.netloc}/{o.get('Prefix')}"
 
     def local_get_folders(self, path: str) -> Iterable[str]:
         if not os.path.isdir(path):
             raise FileNotFoundError(
                 f"{path} does not exist or is not a directory. Please check base_path configuration."
             )
-        for _, folders, _ in os.walk(path):
-            for folder in folders:
-                yield folder
-            break
```

Comment:
Pretty sure `listdir` is just a simpler way to do this

Reply:
i believe os.walk is recursive whereas listdir is not?

Reply:
Yeah but we break after the first loop lol

Reply:
oh huh i guess you're right then